app/webpack/observations/uploader/components/time_shifter.jsx

```diff
@@ -0,0 +1,171 @@
+    const { inputFormat } = this.props;
+
+    cardsToUpdate.forEach( card => {
+      const { date } = card;
+      if ( date === null ) { return; }
+
+      const newDate = moment( new Date( date ) )
+        .add( hours, "hours" )
+        .add( minutes, "minutes" );
+      const currentTime = moment( new Date( ) );
+
+      const minDateString = moment.min( currentTime, newDate );
+      const dateString = newDate.tz( card.time_zone ).format( inputFormat || "YYYY/MM/DD h:mm A ZZ" );
+
+      if ( minDateString !== currentTime ) {
+        this.updateCard( card, dateString );
+      }
+    } );
+  }
+
+  subtractTimeFromSelectedObs( hours, minutes, cardsToUpdate ) {
+    const { inputFormat } = this.props;
+
+    cardsToUpdate.forEach( card => {
+      const { date } = card;
+      if ( date === null ) { return; }
+
+      const dateString = moment( new Date( date ) )
+        .tz( card.time_zone )
+        .subtract( hours, "hours" )
+        .subtract( minutes, "minutes" )
+        .format( inputFormat || "YYYY/MM/DD h:mm A ZZ" );
+
+      this.updateCard( card, dateString );
+    } );
+  }
+
+  handleValueChange( ) {
+    const { selectedObsCards } = this.props;
+    const cardsToUpdate = _.keys( selectedObsCards ).map( card => selectedObsCards[card] );
+
+    const isPositive = value => Math.sign( value ) === 1;
+    const isNegative = value => Math.sign( value ) === -1;
+
+    const diff = ( a, b ) => {
+      if ( a > b ) {
+        return a - b;
+      }
+      return b - a;
+    };
+
+    const {
+      timeShift: value,
+      prevTimeShift
+    } = this.state;
+
+    const amountToShift = value > prevTimeShift
+      ? diff( value, prevTimeShift )
+      : -diff( value, prevTimeShift );
+
+    this.setState( { prevTimeShift: value } );
+
+    const hours = Math.abs( Math.trunc( amountToShift ) );
+    const minutes = Number.isInteger( amountToShift ) ? 0 : 30;
+
+    if ( isPositive( amountToShift ) ) {
+      this.addTimeToSelectedObs( hours, minutes, cardsToUpdate );
+    }
+
+    if ( isNegative( amountToShift ) ) {
+      this.subtractTimeFromSelectedObs( hours, minutes, cardsToUpdate );
+    }
+  }
+
+  handleSlider( ) {
+    // adapted from https://www.w3schools.com/howto/howto_js_rangeslider.asp
+    const slider = document.getElementById( "timeShifter" );
+    const output = document.getElementById( "newTime" );
+    const { value } = slider;
+    output.innerHTML = value; // Display the default slider value
+
+    this.setState( { timeShift: Number( value ) }, ( ) => this.handleValueChange( ) );
+
+    // Update the current slider value (each time you drag the slider handle)
+    slider.oninput = ( ) => {
+      output.innerHTML = this.value;
+    };
+  }
+
+  render( ) {
+    const { timeShift } = this.state;
+    return (
+      <div className="slider-group">
+        <p className="panel-group current-hours">
+          {I18n.t( "hours_adjusted" )}
+          <span id="newTime" />
```

Comment:
I generally try to avoid the use of `id` attributes in react components like this. The `id` is supposed to be unique across the entire page, but a page might have multiple instances of a component. If that occurs, code like `document.getElementById( "newTime" );` might not get the element you think it should.

What I've been doing instead is to add a component specific class to the root element of the component, like `TimeShifter` (camel-case indicates to me that this is the root element of a React component of the same name), and use hyphenated class names for any child elements, like `new-time`. In CSS I can refer to them in a component-specific manner like `.TimeShifter .new-time`.

Accessing the element in JS is a bit trickier. The not-so-great way I've done a bunch is to use ReactDOM and jQuery like this:

```javascript
const domNode = ReactDOM.findDOMNode( this );
const output = $( ".new-time", domNode ).get( 0 ); // get the first element with class .new-time inside of domNode
```

That way I can be sure the element I'm working with is the one inside of this specific instance of this React component. But, as noted at https://reactjs.org/docs/react-dom.html#finddomnode, `ReactDOM.findDOMNode` is not the preferred way of doing this. Instead, we're supposed to use a ref:

```html
<span className="new-time" ref="newTime" />
```
```javascript
const output = this.newTime.current;
// Depending on where you're calling this in the component lifecycle,
// the element might not yet exist in the DOM
if (output) {
  // do something with the element
}
```

None of that matters that much here since there's only one time shifter on the page, but it's a good general practice to avoid bugs in the future.